lisa/tests/scheduler/misfit.py

```diff
@@ -66,16 +66,6 @@ def _get_max_lb_interval(cls, plat_info):
         HZ = plat_info['kernel']['config']['CONFIG_HZ']
         return ((HZ * plat_info['cpus-count']) // 10) * (1. / HZ)
 
-    @classmethod
-    def from_target(cls, target:Target, res_dir:ArtifactPath=None, ftrace_coll:FtraceCollector=None) -> 'MisfitMigrationBase':
```

Comment:
How does that work? How do we end up with something annotated as returning a misfit-typed bundle?

Reply:
`TestBundleMeta` documents what happens:
1) If `_from_target` is defined in a class but `from_target` is not, a dummy `from_target` is created, just calling `super().from_target()`
2) The annotations of `_from_target` are moved to the dummy `from_target`. 

This means that having an annotated `_from_target` is now enough, there is no need for the boilerplate anymore. Since annotations are removed from `_from_target`, it will not be picked up by exekall. 

All in all, that is almost completely transparent to the user, no "weird" behavior should happen. It's still observable, since definitions of `from_target` will now pop, and the name of the method will reflect that it was redefined in the subclass (which is exactly what we want here).

Reply:
Note that in Python 3.6, all the uses of metaclasses we have in `lisa` will be convertible to `__init_subclass__`:

https://docs.python.org/3/reference/datamodel.html#customizing-class-creation

Reply:
That's my point, there is no `_from_target()` in `misfit.py`. Does `RTATestBundle`'s new `from_target()` suffice? It's only annotated as returning a `RTATestBundle`, although I must admit I'm a bit confused since it looks like for misfit we only had something annotated as returning a `MisfitMigrationBase` while the only concrete implementation was `StaggeredFinishes`

Reply:
Ah sorry, missed your point. `RTATestBundle.from_target()` is enough indeed, as misfit was not adding extra parameters. `exekall` is aware of the subclass hierarchy and uses it to create its expressions. What is detected as factory methods (classmethod with return annotation == the class it is defined in) have a "special" treatment: `exekall` considers them as if they were defined on each subclass, returning an instance of that subclass. As long as all the intermediate classes are abstract or don't have any test methods (returning `ResultBundle`), they will not be picked up.

Reply:
Okay, that clears up some of my confusion, thanks.